File: khs/삼각김밥.cpp

#include<iostream>
#include <unordered_map>
#include <string.h>

using namespace std;

#define MAX_N  10001
#define MAX_LEN  141

typedef pair<int, int> P;

unordered_map<int, P> pos;

int main(int argc, char** argv)
{	
    ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	int test_case;
	int T;
    
	
	cin>>T;

    int graph[MAX_LEN][MAX_LEN];

    memset(graph, -1, sizeof(graph));

    int cnt, idx, num;
    cnt = 0;
    idx = 0;
    num = 1;
    while(1){
        graph[idx][cnt] = num;

        pos[num] = P(idx, cnt);
        cnt += 1;
        num += 1;
        if (idx + 1 ==cnt){
            idx += 1;
            cnt = 0;
        }
        if (num == MAX_N)
            break;
    }
    int s, e, x, y;
    int x_0, y_0, x_1, y_1, x_dis, y_dis;
	for(test_case = 1; test_case <= T; ++test_case)
	{
        cin >> s >> e;
        // s = 1;
        // e = 10000;
        cnt = 0;
        if (s > e){
            int temp = e;
            e = s;
            s = temp;
        }
        x_0 = pos[s].first;
        y_0 = pos[s].second;
        x_1 = pos[e].first;
        y_1 = pos[e].second;
        x_dis = x_1 - x_0;
        y_dis = y_1 - y_0;

        if (x_dis > 0 && y_dis > 0) {
            int fast_cnt = min(x_dis, y_dis);
            cnt += fast_cnt;
            x_dis -= fast_cnt;
            y_dis -= fast_cnt;
        }       
        cout << "#" << test_case << " "<< cnt + abs(x_dis) + abs(y_dis) << "\n";

		/////////////////////////////////////////////////////////////////////////////////////////////
		/*
			 이 부분에 여러분의 알고리즘 구현이 들어갑니다.
		 */
		/////////////////////////////////////////////////////////////////////////////////////////////


	}
	return 0;//정상종료시 반드시 0을 리턴해야합니다.
}

File: khs/삼각김밥.py
from collections import defaultdict, deque
dict = defaultdict
pos = dict(int)
dis = dict(int)



dx = [-1, 1, 0, 0,-1, 1]
dy = [ 0, 0,-1, 1,-1, 1]

MAX_N = 10001
MAX_LEN = 141


graph = [[-1] * MAX_LEN for _ in range(MAX_LEN)]
check = [[0] * MAX_LEN for _ in range(MAX_LEN)]


cnt = 0
idx = 0
num = 1
while True:
    graph[idx][cnt] = num
    check[idx][cnt] = 0
    pos[num] = (idx, cnt)
    cnt += 1
    num += 1
    if idx + 1 ==cnt:
        idx += 1
        cnt = 0
    if num == MAX_N:
        break
T = int(input())

for t in range(1, T+1):
    s, e = map(int, input().split())

    q = deque()
    
    x, y = pos[s]
    q.append((x, y, 0))
    check[x][y] = t
    while q:
        x, y, cnt = q.popleft()
        if graph[x][y] == e:
            print(cnt)
            break
        
        
        for i in range(6):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or nx >= MAX_LEN or ny < 0 or ny >= MAX_LEN:
                continue
            if graph[nx][ny] == 0:
                continue
            if check[nx][ny] == t:
                continue
            check[nx][ny] = t
            q.append((nx, ny, cnt + 1))